pageserver/src/storage_sync.rs

```diff
@@ -129,1441 +114,661 @@
-                Ok(()) => {
-                    let mut g = index.write().await;
-
-                    match g.set_awaits_download(&sync_id, false) {
-                        Ok(()) => {
-                            let timeline = g
-                                .timeline_entry_mut(&sync_id)
-                                .expect("set_awaits_download verified existence");
-
-                            timeline.merge_metadata_from_downloaded(
-                                &download_data.data.gathered_metadata,
-                            );
-
-                            register_sync_status(sync_id, sync_start, TASK_NAME, Some(true));
-                            return DownloadStatus::Downloaded;
-                        }
-                        Err(e) => {
-                            error!("Timeline {sync_id} was expected to be in the remote index after a successful download, but it's absent: {e:?}");
-                        }
-                    };
-                }
-                Err(e) => {
-                    error!("Failed to update local timeline metadata: {e:?}");
-                    download_data.retries += 1;
-                    sync_queue.push(sync_id, SyncTask::Download(download_data));
-                    register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
-                }
-            }
-        }
+        info!("scheduled layer file upload {}", path.display());
+
+        // Launch the task immediately, if possible
+        self.launch_queued_tasks(upload_queue);
+        Ok(())
     }
 
-    DownloadStatus::Nothing
-}
+    ///
+    /// Launch a delete operation in the background.
+    ///
+    /// The deletion won't actually be performed, until all preceding
+    /// upload operations have completed succesfully.
+    pub fn schedule_layer_file_deletion(self: &Arc<Self>, paths: &[PathBuf]) -> anyhow::Result<()> {
+        let mut upload_queue = self.upload_queue.lock().unwrap();
+        ensure!(
+            upload_queue.latest_metadata.is_some(),
+            "upload queue not initialized"
+        );
 
-async fn update_local_metadata(
-    conf: &'static PageServerConf,
-    sync_id: TenantTimelineId,
-    remote_timeline: Option<&RemoteTimeline>,
-) -> anyhow::Result<()> {
-    let remote_metadata = match remote_timeline {
-        Some(timeline) => &timeline.metadata,
-        None => {
-            debug!("No remote timeline to update local metadata from, skipping the update");
-            return Ok(());
+        // Update the remote index file, removing the to-be-deleted files from the index,
+        // before deleting the actual files.
+        for path in paths {
+            let relative_path = RelativePath::from_local_path(
+                &self.conf.timeline_path(&self.timeline_id, &self.tenant_id),
+                path,
+            )?;
+            upload_queue.latest_files.remove(&relative_path);
```

Comment:
I still don't fully understand the importance of `latest_files`, but if we want it to be consistent, we need to add `relative_path` back to it at every `?` in this function. We don't currently.

Reply:
@hlinnaka this one is still relevant.

To elaborate:
If we delete multiple paths, and we're in say the second iteration of the loop.
Then we've already removed one path without scheduling the deletion, because we do that after index upload.

If now in the second iteration we return because of `RelativePath::from_local_path(...)?;`, then we've updates `latest_files` without scheduling the deletion.
Not good.

Further, since we return an anyhow::Result, the caller can't distinguish which paths have been queued vs not.

Let's make this atomic. Either all removals are scheduled, or none.

Reply:
Oh I see. It's pretty theoretical that this function could fail after the `upload_queue.latest_files.remove(&relative_path);` calls, but fair enough, let's tighten it up even more. I'll do that.

Reply:
Opened https://github.com/neondatabase/neon/pull/2936 for that